File: AncientVisualResearchSuite/mobile/android/app/src/main/java/com/amerhwitat/avrs/HistoricalScene.kt
package com.amerhwitat.avrs

data class Evidence(val id:String,val source:String,val kind:String="observed",val confidence:Double=1.0)
data class CharacterAgent(val id:String,val role:String,var x:Double,var y:Double,val heading:Double=0.0,val activity:String="idle")
data class HistoricalScene(val id:String,val title:String,val latitude:Double,val longitude:Double,val evidence:List<Evidence>,val characters:MutableList<CharacterAgent>) {
    fun simulate(seconds:Double, step:Double=.25):List<List<CharacterAgent>> {
        require(seconds>=0 && step>0)
        val frames=mutableListOf<List<CharacterAgent>>()
        var t=0.0
        while(t<=seconds+1e-9){
            frames += characters.map { c ->
                val active=if(c.activity=="idle"||c.activity=="sleep") 0.0 else .5
                val phase=Math.toRadians(c.heading)+t*.15
                c.copy(x=c.x+kotlin.math.cos(phase)*active*t,y=c.y+kotlin.math.sin(phase)*active*t)
            }
            t+=step
        }
        return frames
    }
}

File: AncientVisualResearchSuite/mobile/ios/HistoricalScene.swift
import Foundation

public struct Evidence: Codable { public let id:String; public let source:String; public let kind:String; public let confidence:Double }
public struct CharacterAgent: Codable { public let id:String; public let role:String; public var x:Double; public var y:Double; public let heading:Double; public let activity:String }
public struct HistoricalScene: Codable {
    public let id:String; public let title:String; public let latitude:Double; public let longitude:Double
    public var characters:[CharacterAgent]; public var evidence:[Evidence]
    public func simulate(seconds:Double, step:Double=0.25) -> [[CharacterAgent]] {
        precondition(seconds >= 0 && step > 0)
        var frames:[[CharacterAgent]]=[]; var t=0.0
        while t <= seconds + 1e-9 {
            frames.append(characters.map { c in
                let active=(c.activity == "idle" || c.activity == "sleep") ? 0.0 : 0.5
                let phase=c.heading * .pi / 180.0 + t * 0.15
                return CharacterAgent(id:c.id,role:c.role,x:c.x + cos(phase)*active*t,y:c.y + sin(phase)*active*t,heading:c.heading,activity:c.activity)
            })
            t += step
        }
        return frames
    }
}
